Add "editor" extension type

zjpiazza opened the pull request:
Add a new 'editor' extension type that allows developers to create custom item editing interfaces that can be selected per collection, following the same patterns as existing extension types (layouts, interfaces, displays).

## Core Implementation

### Type System & Configuration
- Add complete TypeScript definitions for editor extensions
- Add EditorConfig interface with id, name, icon, component properties
- Add EditorProps interface defining collection, primaryKey, item, v-model support
- Integrate editors into APP_EXTENSION_TYPES and extension system

### Extension Registration & Management
- Add defineEditor() function following established patterns
- Implement editor extension registration and loading
- Add internal editor management system

### User Interface Integration
- Add editor selection sidebar component for collection item pages
- Implement conditional rendering of custom editor components in item view
- Add editor preset support for persistent per-collection selection
- Add database migration for editor field in directus_presets table

## Usage

```typescript
import { defineEditor } from '@directus/extensions-sdk';

export default defineEditor({
  id: 'my-custom-editor',
  name: 'My Custom Editor', 
  icon: 'edit',
  component: MyEditorComponent,
});
```

Editor components receive collection, primaryKey, item props and support v-model for edits.

## Implementation Details

- **14 files changed**: Core extension system integration
- **Database migration**: Adds `editor` field to `directus_presets` table
- **Preset persistence**: Editor selection saved per collection like layout preferences
- **Conditional rendering**: Seamless fallback to default form when no editor selected
- **Extension loading**: Auto-discovery and registration of editor extensions

## Testing

- ✅ **Unit tests** for `defineEditor` function (added to existing test suite)
- ✅ **Extension registration** verified through build process
- ✅ **TypeScript compilation** and type safety validation
- ✅ **Database migration** runs without errors
- ✅ **UI component rendering** manually verified
- ✅ **Preset persistence** tested across browser sessions
- ✅ **Backward compatibility** with existing collections confirmed

**Test Coverage:**
- Added test case for `defineEditor()` in `define-extension.test.ts` 
- Validated editor config object structure and return values
- All existing extension system tests continue to pass

## Review Notes

- This follows the exact same architectural patterns as existing extension types
- No breaking changes - purely additive functionality
- Migration is safe and only adds a nullable column

## Checklist

- [x] **Added or updated tests** - Added unit test for `defineEditor` function
- [ ] **Documentation PR created** [here](https://github.com/directus/docs) or not required

zjpiazza committed:
feat: implement editor extension type for custom item editing interfaces

Add a new 'editor' extension type that allows developers to create custom item editing interfaces that can be selected per collection, following the same patterns as existing extension types (layouts, interfaces, displays).

## Core Implementation

### Type System & Configuration
- Add complete TypeScript definitions for editor extensions
- Add EditorConfig interface with id, name, icon, component properties
- Add EditorProps interface defining collection, primaryKey, item, v-model support
- Integrate editors into APP_EXTENSION_TYPES and extension system

### Extension Registration & Management
- Add defineEditor() function following established patterns
- Implement editor extension registration and loading
- Add internal editor management system

### User Interface Integration
- Add editor selection sidebar component for collection item pages
- Implement conditional rendering of custom editor components in item view
- Add editor preset support for persistent per-collection selection
- Add database migration for editor field in directus_presets table

### Developer Experience
- Complete TypeScript support and intellisense
- Follows exact same patterns as layout/interface/display extensions
- Backward compatible with existing functionality

## Usage

Editor components receive collection, primaryKey, item props and support v-model for edits.

zjpiazza committed:
feat: implement editor extension type for custom item editing interfaces

Add a new 'editor' extension type that allows developers to create custom item editing interfaces that can be selected per collection, following the same patterns as existing extension types (layouts, interfaces, displays).

## Core Implementation

### Type System & Configuration
- Add complete TypeScript definitions for editor extensions
- Add EditorConfig interface with id, name, icon, component properties
- Add EditorProps interface defining collection, primaryKey, item, v-model support
- Integrate editors into APP_EXTENSION_TYPES and extension system

### Extension Registration & Management
- Add defineEditor() function following established patterns
- Implement editor extension registration and loading
- Add internal editor management system

### User Interface Integration
- Add editor selection sidebar component for collection item pages
- Implement conditional rendering of custom editor components in item view
- Add editor preset support for persistent per-collection selection
- Add database migration for editor field in directus_presets table

### Developer Experience
- Complete TypeScript support and intellisense
- Follows exact same patterns as layout/interface/display extensions
- Backward compatible with existing functionality

## Usage

Editor components receive collection, primaryKey, item props and support v-model for edits.

zjpiazza committed:
Add zjpiazza to contributors list

zjpiazza committed:
Fix failing checks

* Add changeset
* Format
* Lint

zjpiazza committed:
Merge branch 'main' into editor-extension-type

zjpiazza committed:
fix: update extension entrypoint test snapshots for editor extension type

Updated inline snapshots in generateExtensionsEntrypoint tests to include the new 'editors' export that was added as part of the editor extension type implementation.

zjpiazza committed:
test: add test coverage for editor extension functionality

- Add unit tests for editors/index.ts covering registerEditors and deregisterEditors functions
- Add Vue component tests for editor-sidebar-detail.vue covering component rendering and event emission
- Improve test coverage for editor extension type implementation to meet codecov targets

zjpiazza committed:
Add migration test coverage and fix formatting

JoshTheDerf commented:
Is there an existing discussion/issue this ties into? If so, great! If not, you might need to get that going first before this could be considered for merging.

I personally would love this functionality. Right now it can be accomplished by implementing a custom module that overrides the /collection/ID route, but it's hacky, and requires reimplementing everything in the right sidebar.

zjpiazza commented:
> Is there an existing discussion/issue this ties into? If so, great! If not, you might need to get that going first before this could be considered for merging.
> 
> I personally would love this functionality. Right now it can be accomplished by implementing a custom module that overrides the /collection/ID route, but it's hacky and requires reimplementing everything in the right sidebar.

Hey Josh! There is no conversation regarding this feature yet, but that was my intention with this PR. I am currently working on a Directus implementation with a strong need for this so regardless if the PR is accepted or not, I'll be deploying from my forked version.

I am very new to  working with Directus and I understand that extensions are a big architectural pillar of the software so its quite possible my implementation here has  some serious gaps.

Open to any and all thoughts on this.

zjpiazza committed:
Added new extension type templates to `extensions-sdk`

zjpiazza committed:
fix: prevent undefined error in ExtensionsService bundled extensions processing

- Add null check for parentBundle.schema before accessing entries
- Add optional chaining to entries.find() for defensive programming
- Fixes error: Cannot read properties of undefined (reading 'find')

zjpiazza committed:
Add refresh event hook to editor component
